samples/performance/async_compute/async_compute.cpp

```diff
@@ -0,0 +1,809 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "async_compute.h"
+
+#include "api_vulkan_sample.h"
+#include "common/vk_common.h"
+#include "gltf_loader.h"
+#include "gui.h"
+#include "platform/filesystem.h"
+#include "platform/platform.h"
+#include "scene_graph/components/orthographic_camera.h"
+#include "stats/stats.h"
+
+AsyncComputeSample::AsyncComputeSample()
+{
+	auto &config = get_configuration();
+
+	config.insert<vkb::BoolSetting>(0, async_enabled, false);
+	config.insert<vkb::BoolSetting>(1, async_enabled, true);
+	config.insert<vkb::BoolSetting>(0, rotate_shadows, false);
+	config.insert<vkb::BoolSetting>(1, rotate_shadows, true);
+	config.insert<vkb::BoolSetting>(0, double_buffer_hdr_frames, false);
+	config.insert<vkb::BoolSetting>(1, double_buffer_hdr_frames, true);
+}
+
+void AsyncComputeSample::draw_gui()
+{
+	gui->show_options_window(
+	    /* body = */ [this]() {
+		    ImGui::Checkbox("Enable async queues", &async_enabled);
+		    ImGui::Checkbox("Double buffer HDR", &double_buffer_hdr_frames);
+		    ImGui::Checkbox("Rotate shadows", &rotate_shadows);
+	    },
+	    /* lines = */ 3);
+}
+
+static VkExtent3D downsample_extent(const VkExtent3D &extent, uint32_t level)
+{
+	return {
+	    std::max(1u, extent.width >> level),
+	    std::max(1u, extent.height >> level),
+	    std::max(1u, extent.depth >> level)};
+}
+
+void AsyncComputeSample::prepare_render_targets()
+{
+	// To make this sample demanding enough to saturate mobile devices, use 4K.
+	// Could base this off the swapchain extent, but comparing cross-device performance
+	// could get awkward.
+	VkExtent3D size = {3840, 2160, 1};
+
+	// Support double-buffered HDR.
+	vkb::core::Image color_targets[2]{
+	    {*device, size, VK_FORMAT_R16G16B16A16_SFLOAT,
+	     VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT,
+	     VMA_MEMORY_USAGE_GPU_ONLY},
+	    {*device, size, VK_FORMAT_R16G16B16A16_SFLOAT,
+	     VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT,
+	     VMA_MEMORY_USAGE_GPU_ONLY},
+	};
+
+	// Should only really need one depth target, but vkb::RenderTarget needs to own the resource.
+	vkb::core::Image depth_targets[2]{
+	    {*device, size, VK_FORMAT_D32_SFLOAT,
+	     VK_IMAGE_USAGE_TRANSIENT_ATTACHMENT_BIT | VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT,
+	     VMA_MEMORY_USAGE_GPU_ONLY},
+	    {*device, size, VK_FORMAT_D32_SFLOAT,
+	     VK_IMAGE_USAGE_TRANSIENT_ATTACHMENT_BIT | VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT,
+	     VMA_MEMORY_USAGE_GPU_ONLY},
+	};
+
+	// 8K shadow-map overkill to stress devices.
+	vkb::core::Image shadow_target{*device, VkExtent3D{8 * 1024, 8 * 1024, 1}, VK_FORMAT_D16_UNORM,
```

Comment:
8K images aren't universally supported by Vulkan (or some of our devices).  You should be using 4K maximum for guaranteed support, or checking the limits and selecting the best-fit power-of-two size perhaps.

Reply:
Fixed.

Reply:
Looks good, thanks.